### DISCUSSION TITLE
GitHub Actions Numeric Comparison Operators Are Incorrect

### DISCUSSION THREAD

**Discussion by norwd** [community/community#57480] [Actions]
### Select Topic Area

Bug

### Body

I have a workflow that uses an `if: steps.foo.outputs.blah < steps.bar.outputs.blah` condition to only run a step if one calculated value is greater than another calculated value. This steps is getting run when it shouldn't be and as far as I can tell this is a bug in the GitHub Actions Expressions.

The debug output shows that the condition being evaluated was `15 < 9`, which should be `false` but was evaluated as `true`.

```
##[debug]Expanded: (true && ('15' < '9'))
##[debug]Result: true
```

My theory is that this is because the step outputs here are being evaluated as strings instead of as numbers. As an experiment I tried using the `fromJSON()` function to evaluate the strings as numbers, since `"15"` is valid JSON for the number 15, and this appears to work correctly. The [documentation](https://docs.github.com/en/actions/learn-github-actions/expressions#operators) for the expressions in GitHub Actions doesn't clarify this and I think this caveat should be at least mentioned if not fixed outright.

TL;DR: `${{ steps.foo.outputs.blah < steps.bar.outputs.blah }}` is wrong, use `${{ fromJSON(steps.foo.outputs.blah) < fromJSON(steps.bar.outputs.blah) }}` instead as a workaround.

**Comment by norwd** [comment]
Note: I've proposed updating the documentation of this behaviour in https://github.com/github/docs/pull/25870

**Comment by ghostinhershell** [comment]
Hey @norwd 👋 

This sounds like it could be helpful for the whole GitHub community, thank you for opening a PR to get this updated! Our docs are open sourced so everyone in our community is welcome to [contribute](https://github.com/github/docs/blob/main/CONTRIBUTING.md#welcome-to-github-docs-contributing-guide-) and help users all over the world. Thanks again!
